comps/llms/rayllm/api_server_openai.py

```diff
@@ -0,0 +1,137 @@
+
+import ray
+from easydict import EasyDict as edict
+from ray import serve
+from rayllm.api_openai_backend.query_client import RouterQueryClient
+from rayllm.api_openai_backend.router_app import Router, router_app
+from rayllm.ray_serve import LLMServe
+
+
+def router_application(deployments, max_concurrent_queries):
+    """Create a Router Deployment.
+
+    Router Deployment will point to a Serve Deployment for each specified base model,
+    and have a client to query each one.
+    """
+    merged_client = RouterQueryClient(deployments)
+
+    RouterDeployment = serve.deployment(
+        route_prefix="/",
+        max_concurrent_queries=max_concurrent_queries,  # Maximum backlog for a single replica
+    )(serve.ingress(router_app)(Router))
+
+    return RouterDeployment.bind(merged_client)
+
+
+def openai_serve_run(deployments, host, route_prefix, port, max_concurrent_queries):
+    router_app = router_application(deployments, max_concurrent_queries)
+
+    serve.start(http_options={"host": host, "port": port})
+    serve.run(
+        router_app,
+        name="router",
+        route_prefix=route_prefix,
+    ).options(
+        stream=True,
+        use_new_handle_api=True,
+    )
+    deployment_address = f"http://{host}:{port}{route_prefix}"
+    print(f"Deployment is ready at `{deployment_address}`.")
+    return deployment_address
+
+
+def get_deployment_actor_options(hpus_per_worker, ipex_enabled=False):
+    _ray_env_key = "env_vars"
+    # OMP_NUM_THREADS will be set by num_cpus, so not set in env
+    _predictor_runtime_env_ipex = {
+        "KMP_BLOCKTIME": "1",
+        "KMP_SETTINGS": "1",
+        "KMP_AFFINITY": "granularity=fine,compact,1,0",
+        "MALLOC_CONF": "oversize_threshold:1,background_thread:true,\
+            metadata_thp:auto,dirty_decay_ms:9000000000,muzzy_decay_ms:9000000000",
+    }
+    runtime_env: Dict[str, Any] = {_ray_env_key: {}}
+    if ipex_enabled:
+        runtime_env[_ray_env_key].update(_predictor_runtime_env_ipex)
+    ray_actor_options: Dict[str, Any] = {"runtime_env": runtime_env}
+    ray_actor_options["resources"] = {"HPU": hpus_per_worker}
+
+    return ray_actor_options
+
+
+def main(argv=None):
+    import argparse
+
+    parser = argparse.ArgumentParser(description="Serve a router for models on Ray.", add_help=True)
+    parser.add_argument("--port_number", default=8080, type=int, help="Port number to serve on.")
+    parser.add_argument(
+        "--model_id_or_path", default="meta-llama/Llama-2-7b-chat-hf", type=str, help="Model id or path."
+    )
+    parser.add_argument("--max_num_seqs", default=256, type=int, help="Maximum number of sequences to generate.")
+    parser.add_argument("--max_batch_size", default=8, type=int, help="Maximum batch size.")
+    parser.add_argument("--num_replicas", default=1, type=int, help="Number of replicas to start.")
+    parser.add_argument("--num_workers_per_group", default=8, type=int, help="Number of workers per group.")
```

Comment:
Those variables are only effective when Intel OpenMP and jemalloc are installed. Can consider remove these.

Reply:
Removed "num_workers_per_group" parameter. Thanks.